### PR TITLE
[ROCm] HIP stream priority fix post #101956

### PR DESCRIPTION
**Pull request by jeffdaily**
PR #101956 introduced additional stream priorities for cuda streams. HIP streams have slightly different semantics.
- HIP: 1=low, 0=default, -1=high
- CUDA: 0=default, -1=high, -2=higher, etc.

This PR forces HIP stream priority to just 0 and -1 to match the pytorch semantics.

This fixes a broken unit test.

```
python3 test_cuda_multigpu.py TestCudaMultiGPU.test_streams_priority -v

Test results will be stored in test-reports/python-unittest/test_cuda_multigpu

Running tests...
----------------------------------------------------------------------
  test_streams_priority (__main__.TestCudaMultiGPU) ... ERROR (0.200s)

======================================================================
ERROR [0.200s]: test_streams_priority (__main__.TestCudaMultiGPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_utils.py", line 2354, in wrapper
    method(*args, **kwargs)
  File "test_cuda_multigpu.py", line 656, in test_streams_priority
    low, high = torch.cuda.Stream.priority_range()
RuntimeError: least_priority == 0 INTERNAL ASSERT FAILED at "/var/lib/jenkins/pytorch-upstream/c10/hip/HIPStream.h":184, please report a bug to PyTorch. Unexpected HIP stream priority range
```

cc @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

File: c10/cuda/CUDAStream.cpp [hunk 1]
```diff
@@ -163,6 +163,11 @@ static void initGlobalStreamState() {
   int leastPriority = -1, greatestPriority = -1;
   C10_CUDA_CHECK(
       cudaDeviceGetStreamPriorityRange(&leastPriority, &greatestPriority));
+  // Note [HIP stream priorities]
+  // HIP stream priorities are 1=low, 0=default, -1=high which differs from CUDA
+  // which is 0=default, -1=high, -2=higher etc.
+  // Clamp leastPriority to 0 for HIP.
+  leastPriority = 0;
```
Comment (malfet) [inline c10/cuda/CUDAStream.cpp:170]: Shouldn't this be guarded with `USE_ROCM`?

Comment (malfet) [reply c10/cuda/CUDAStream.cpp:170]: Also, are you saying that `cudaDeviceGetStreamPriorityRange` is not clamping priorities correctly on ROCM?

Comment (jeffdaily) [reply c10/cuda/CUDAStream.cpp:170]: Yeah, I debated needing an ifdef there.  cuda returns 0 for leastPriority, so it's a no-op for cuda.  hip returns 1.  it's a behavior difference between HIP and CUDA (see Note I added in comments in the PR diff).  HIP and CUDA agree that 0 is the default stream priority.  but HIP adds a lower priority value of 1 in addition to the higher priority value of -1.  This change here is to make HIP behave in accordance with the semantics of pytorch and cuda where 0 is the lowest AND default priority.  We're basically ignoring the fact that HIP provides a lower priority than 0.  Later in the code, the priority range is iterated over and without this change HIP would end up creating 3 stream pools for priorities 0, -1, and another -1.

I don't expect CUDA will ever have leastPriority NOT be 0, and there is a TORCH_ASSERT elsewhere in the code to make sure it never changes.  Do you still want this #ifdef protected for USE_ROCM?

Comment (malfet) [reply c10/cuda/CUDAStream.cpp:170]: I would prefer to keep an ifdef here and do not assume what `cudaDeviceGetStreamPriorityRange` returns/might return in the next version of CUDA

Comment (jeffdaily) [inline c10/cuda/CUDAStream.cpp:170]:
```suggestion
#ifdef USE_ROCM
  leastPriority = 0;
#endif
```

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**